### ISSUE TITLE
Process gets killed while calculating similarity for 'jester' dataset

### ISSUE THREAD

**Issue by xahiru** [NicolasHug/Surprise#234]
#### Description

Hi,
I'm trying to run the following codes on my Mac(8GB RAM). When I changed the sim_options ```user_based``` to ```True```, system gets super slow and eventually the process gets killed. So I tried it on my server(with 128GB RAM), but still gets the same problem. However, when ```user_based``` is set to ```False```, it works smoothly. Also, with 'ml-1m' dataset it runs smoothly. Is it still due to insufficient memory?

Thanks

#### Steps/Code to Reproduce

```from surprise import Dataset
from surprise import KNNWithMeans
from surprise.model_selection import train_test_split
from surprise.accuracy import rmse

data = Dataset.load_builtin('jester')
trainset, testset = train_test_split(data, test_size=.2, train_size=None, random_state=100, shuffle=True)
sim_options={'name':'pearson','user_based':True}
algo = KNNWithMeans(sim_options=sim_options,verbose=True)
algo.fit(trainset)
predictions=algo.test(testset)
rmse(predictions)
```
#### Expected Results
RMSE: [some value]
#### Actual Results
Computing the pearson similarity matrix..
Killed

#### Versions

Linux-3.10.0-957.1.3.el7.x86_64-x86_64-with-centos-7.6.1810-Core
Python 3.6.3 (default, Mar 20 2018, 13:50:41)
[GCC 4.8.5 20150623 (Red Hat 4.8.5-16)]
surprise 1.0.6

and

Darwin-18.2.0-x86_64-i386-64bit
Python 3.6.0b4 (v3.6.0b4:18496abdb3d5, Nov 21 2016, 20:44:47)
commit 131564c4857c3e75e0914999bbc32f185dfd301b (HEAD -> master, origin/master, origin/HEAD)
[GCC 4.2.1 (Apple Inc. build 5666) (dot 3)]
surprise 1.0.6

**Comment by NicolasHug** [comment]
>Is it still due to insufficient memory?

Yes most definitely. Jester has about 25k users and only 100 items so the user-based sim takes much more space. It's weird that it doesn't work with your 128GB server though because as far as I can tell, with 25k users all then temporary structures durint the computation shouldn't take more than 16GB.

Maybe you'll want to switch to more scalable algorithms such as SVD?

**Comment by xahiru** [comment]
Thanks, I'll try to figure it out. Actually, I'm trying to experiment with memory-based methods, so SVD is not an option.

**Comment by NicolasHug** [comment]
Did you manage to get it to work @xahiru ?

**Comment by xahiru** [comment]
No, not yet. So I'm using a subset of the dataset.
The problem is the sheer size of following matrices in `similarities.pyx` :
```
    freq = np.zeros((n_x, n_x), np.int)
    prods = np.zeros((n_x, n_x), np.double)
    sqi = np.zeros((n_x, n_x), np.double)
    sqj = np.zeros((n_x, n_x), np.double)
    si = np.zeros((n_x, n_x), np.double)
    sj = np.zeros((n_x, n_x), np.double)
    sim = np.zeros((n_x, n_x), np.double)
```

The size of each matrix is 58791 by 58791. I added extra 56GB to swap, but still it's not enough. I think if I add extra 100GB of memory it may work. To test the memory requirement for a single matrix, I ran the following codes:

>>> `import copy as cp`
>>> `import numpy as np`
>>> `az = np.zeros((58791,58791),np.double)`
>>> `az += 1`
>>> `bz= cp.deepcopy(az)`
>>> `cz= cp.deepcopy(az)`
>>> `df= cp.deepcopy(az)`
>>> `fz= cp.deepcopy(az)`

and the last line resulted in `MemoryError`. So, I think each matrix requires roughly 30GB.
